File: backend-app/src/main/java/tom/projects/monkeylog/service/ProgramService.java
package tom.projects.monkeylog.service;

import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;
import tom.projects.monkeylog.model.workout.Program;
import tom.projects.monkeylog.repository.workout.ProgramRepository;
import tom.projects.monkeylog.security.AuthenticatedUser;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ProgramService {
    private final ProgramRepository programRepository;

    public List<Program> all() {
        return programRepository.findAllByUserIdAndTemplateIsTrue(AuthenticatedUser.getId());
    }

    public Program get(Long id) {
        return programRepository.findById(id)
                .filter(AuthenticatedUser::isOwner)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Program not found"));
    }

    public Program create(Long id) {
        return programRepository.findById(id)
                .filter(AuthenticatedUser::isOwner)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Program not found"));
    }

    public Program start(Long id) {
        return programRepository.findById(id)
                .filter(AuthenticatedUser::isOwner)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Program not found"));
    }
}

File: backend/src/main/kotlin/tom/com/monkeylog/service/ProgramService.kt
package tom.com.monkeylog.service

import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.web.server.ResponseStatusException
import tom.com.monkeylog.dto.program.ProgramCreateRequest
import tom.com.monkeylog.dto.program.ProgramUpdateRequest
import tom.com.monkeylog.dto.program.ProgramWeekCreateRequest
import tom.com.monkeylog.mapper.toEntity
import tom.com.monkeylog.mapper.update
import tom.com.monkeylog.model.workout.Program
import tom.com.monkeylog.model.workout.ProgramWeek
import tom.com.monkeylog.repository.ProgramRepository
import tom.com.monkeylog.repository.ProgramWeekRepository
import tom.com.monkeylog.security.AuthenticatedUser
import java.util.*

@Service
class ProgramService(
    private val programRepository: ProgramRepository,
    private val programWeekRepository: ProgramWeekRepository,
) {
    fun allPrograms() = programRepository.findAllByUserId(AuthenticatedUser.id)

    fun getProgram(id: UUID): Program = programRepository.findById(id)
        .filter(AuthenticatedUser::isResourceOwner)
        .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "Program not found") }

    fun createProgram(programCreateRequest: ProgramCreateRequest): Program {
        val program: Program = programCreateRequest.toEntity()
        program.userId = AuthenticatedUser.id
        return programRepository.save(program)
    }

    fun updateProgram(id: UUID, programUpdateRequest: ProgramUpdateRequest) =
        programRepository.save(getProgram(id).update(programUpdateRequest))

    fun deleteProgram(id: UUID) = programRepository.delete(getProgram(id))

    fun getProgramWeek(id: UUID): ProgramWeek = programWeekRepository.findById(id)
        .filter(AuthenticatedUser::isResourceOwner)
        .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "Program week not found") }

    fun createProgramWeek(id: UUID, programWeekCreateRequest: ProgramWeekCreateRequest) =
        programWeekRepository.save(programWeekCreateRequest.toEntity(getProgram(id)))

    fun deleteProgramWeek(id: UUID) {
        programWeekRepository.delete(getProgramWeek(id))
    }
}
